src/main/java/org/spongepowered/common/service/profile/SpongeProfileResolver.java

```diff
@@ -0,0 +1,260 @@
+            if (profile == null || !profile.isComplete()) {
+                throw new ProfileNotFoundException("Profile: " + profile);
+            }
+            return (GameProfile) profile;
+        }
+
+        protected List<GameProfile> fromNames(List<String> names) throws Exception {
+            final List<GameProfile> profiles = Lists.newArrayList();
+            Set<String> cachedNames = Sets.newHashSet(this.cache.getUsernames());
+            if (this.useCache) {
+                for (int i = 0; i < names.size(); i++) {
+                    GameProfile profile = null;
+                    if (cachedNames.contains(names.get(i).toLowerCase(Locale.ROOT))) {
+                        profile = (GameProfile) this.cache.getGameProfileForUsername(names.get(i));
+                    }
+                    if (profile != null) {
+                        profiles.add(profile);
+                        names.remove(i--);
+                    }
+                }
+            }
+            if (names.isEmpty()) {
+                return profiles;
+            }
+            final ProfileNotFoundException[] thrown = new ProfileNotFoundException[1];
+            this.server.getGameProfileRepository().findProfilesByNames(names.toArray(new String[0]), Agent.MINECRAFT, new ProfileLookupCallback() {
+
+                @Override
+                public void onProfileLookupSucceeded(com.mojang.authlib.GameProfile profile) {
+                    Query.this.cache.addEntry(profile);
+                    if (thrown[0] == null) {
+                        profiles.add((GameProfile) profile);
+                    }
+                }
+
+                @Override
+                public void onProfileLookupFailed(com.mojang.authlib.GameProfile profile, Exception exception) {
+                    thrown[0] = new ProfileNotFoundException("Profile: " + profile, exception);
+                }
+            });
+            this.cache.save();
+            if (thrown[0] != null) {
+                throw thrown[0];
+            }
+            return profiles;
+        }
+
+    }
+
+    private static class SingleQuery extends Query<GameProfile> {
+
+        private final UUID id;
+        private final String name;
+
+        public SingleQuery(UUID uniqueId, boolean useCache) {
+            super(useCache);
+            this.id = uniqueId;
+            this.name = null;
+        }
+
+        public SingleQuery(String name, boolean useCache) {
+            super(useCache);
+            this.name = name;
+            this.id = null;
+        }
+
+        @Override
+        public GameProfile call() throws Exception {
+            if (this.id != null) {
+                return this.fromId(this.id);
+            } else if (this.name != null) {
+                return this.fromNames(Lists.newArrayList(this.name)).get(0);
+            }
+            throw new IllegalStateException("Impossibru");
```

Comment:
Well it literally is impossible to get to this line because either `id` or `name` will not be null but I made it in an `else if` to make it more clear (an explicit null check),